synapse/handlers/room_member.py

```diff
@@ -681,22 +696,74 @@ def do_3pid_invite(
             )
 
     @defer.inlineCallbacks
-    def _lookup_3pid(self, id_server, medium, address):
+    def _lookup_3pid(self, id_server, medium, address, id_access_token=None):
         """Looks up a 3pid in the passed identity server.
 
         Args:
             id_server (str): The server name (including port, if required)
                 of the identity server to use.
             medium (str): The type of the third party identifier (e.g. "email").
             address (str): The third party identifier (e.g. "foo@example.com").
+            id_access_token (str|None): The access token to authenticate to the identity
+                server with
 
         Returns:
-            str: the matrix ID of the 3pid, or None if it is not recognized.
+            str|None: the matrix ID of the 3pid, or None if it is not recognized.
         """
-        if not self._enable_lookup:
-            raise SynapseError(
-                403, "Looking up third-party identifiers is denied from this server"
+        # If an access token is present, add it to the query params of the hash_details request
+        query_params = {}
+        if id_access_token is not None:
+            query_params["id_access_token"] = id_access_token
+
+        # Check what hashing details are supported by this identity server
+        use_v1 = False
+        hash_details = None
+        try:
+            hash_details = yield self.simple_http_client.get_json(
+                "%s%s/_matrix/identity/v2/hash_details" % (id_server_scheme, id_server),
+                query_params,
             )
+            if not isinstance(hash_details, dict):
+                logger.warn(
+                    "Got non-dict object when checking hash details of %s: %s",
+                    id_server,
+                    hash_details,
+                )
+                raise SynapseError(
```

Comment:
not sure if returning a 500 is the right choice here. Given that the existing code returns None under similar circumstances, shouldn't we do likewise?

(Indeed, I don't think this function has any business deciding what error code will be returned from the API, when we don't even know which endpoint was called.)